fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java

```diff
@@ -0,0 +1,379 @@
+        List<OptExpression> scanExprs = Lists.newArrayList();
+        OptExpressionVisitor scanCollector = new OptExpressionVisitor<Void, Void>() {
+            @Override
+            public Void visit(OptExpression optExpression, Void context) {
+                for (OptExpression input : optExpression.getInputs()) {
+                    super.visit(input, context);
+                }
+                return null;
+            }
+
+            @Override
+            public Void visitLogicalTableScan(OptExpression optExpression, Void context) {
+                scanExprs.add(optExpression);
+                return null;
+            }
+        };
+        expression.getOp().accept(scanCollector, expression, null);
+        return scanExprs;
+    }
+
+    // get all predicates within and below root
+    public static List<ScalarOperator> getAllPredicates(OptExpression root) {
+        List<ScalarOperator> predicates = Lists.newArrayList();
+        getAllPredicates(root, predicates);
+        return predicates;
+    }
+
+    private static void getAllPredicates(OptExpression root, List<ScalarOperator> predicates) {
+        Operator operator = root.getOp();
+        if (operator.getPredicate() != null) {
+            predicates.add(root.getOp().getPredicate());
+        }
+        if (operator instanceof LogicalJoinOperator) {
+            LogicalJoinOperator joinOperator = (LogicalJoinOperator) operator;
+            if (joinOperator.getOnPredicate() != null) {
+                predicates.add(joinOperator.getOnPredicate());
+            }
+        }
+        for (OptExpression child : root.getInputs()) {
+            getAllPredicates(child, predicates);
+        }
+    }
+
+    public static ScalarOperator canonizePredicate(ScalarOperator predicate) {
+        if (predicate == null) {
+            return null;
+        }
+        ScalarOperatorRewriter rewrite = new ScalarOperatorRewriter();
+        return rewrite.rewrite(predicate, ScalarOperatorRewriter.DEFAULT_REWRITE_SCAN_PREDICATE_RULES);
+    }
+
+    public static ScalarOperator canonizePredicateForRewrite(ScalarOperator predicate) {
+        if (predicate == null) {
+            return null;
+        }
+        ScalarOperatorRewriter rewrite = new ScalarOperatorRewriter();
+        return rewrite.rewrite(predicate, ScalarOperatorRewriter.MV_SCALAR_REWRITE_RULES);
+    }
+
+    public static ScalarOperator getCompensationPredicateForDisjunctive(ScalarOperator src, ScalarOperator target) {
+        List<ScalarOperator> srcItems = Utils.extractDisjunctive(src);
+        List<ScalarOperator> targetItems = Utils.extractDisjunctive(target);
+        int srcLength = srcItems.size();
+        int targetLength = targetItems.size();
+        if (!targetItems.containsAll(srcItems)) {
+            return null;
+        }
+        targetItems.removeAll(srcItems);
+        if (targetItems.isEmpty() && srcLength == targetLength) {
```

Comment:
when targetItems and srcItems have no duplicates(because they have been canonized), and targetItems contain all srcItems, so srcItems must be a subset of or equal to targetItems. After removing all srcItems from targetItem, if  targetItems.isEmpty(), it means srcItems are equal to targetItems. if not, it means they are subset.
Here I can simplify the condition.